### PR TITLE
network: introduce multipath route (WIP)

### PR DESCRIPTION
**Pull request by ssahani**
@yuwata could you please take a look at this. Why this is not working.

https://github.com/systemd/systemd/issues/12541

cc @n0rad 
```
$ ip route add default nexthop via 192.168.11.1 dev inet.11 weight 5 nexthop via 192.168.22.2 dev inet.22 weight 10
```

```
[MultiPathRoute]
Gateway=192.168.3.1
Weight=10
Ifname=eth0


[MultiPathRoute]
Gateway=192.168.3.2
Weight=4
Ifname=eth1

```

### CODE REVIEW

**Comment by ssahani** [comment]
@yuwata To give you a hint we are forming the same message same data as iproute. Cant be seen via strace . The kernel does not accept it says `-EINVAL` on netlink messages.

File: src/network/networkd-link.c [hunk 1]
```diff
@@ -1089,6 +1090,9 @@ void link_check_ready(Link *link) {
                 return;
         }
 
+        if (ordered_set_size(link->network->multipath_routes))
```
Comment (yuwata) [inline src/network/networkd-link.c]: `!ordered_set_isempty()`

File: src/network/networkd-multipath-route.c [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+#include "strxcpyx.h"
+#include "sysctl-util.h"
+#include "util.h"
+
+static void multipath_route_hash_func(const MultiPathRoute *id, struct siphash *state) {
+         assert(id);
+         assert(&id->ifindex);
+         assert(&id->family);
+         assert(&id->weight);
+         assert(&id->gw);
```
Comment (yuwata) [inline src/network/networkd-multipath-route.c]: Spurious assertions. Please drop them.

File: src/network/networkd-multipath-route.c [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+        r = ordered_set_ensure_allocated(&route->network->multipath_routes, &multipath_route_hash_ops);
+        if (r < 0)
+                return r;
+
+        r = ordered_set_put(route->network->multipath_routes, route);
+        if (r < 0)
+                return r;
+
+        return 0;
+}
```
Comment (yuwata) [inline src/network/networkd-multipath-route.c]: The naming of this function is misleading...

File: src/network/networkd-multipath-route.c [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+                if (in_addr_is_null(route->family, &route->gw) == 0) {
+                        if (route->family == AF_INET) {
+                                r = rta_add_attribute(rta, RTA_GATEWAY, &route->gw.in, FAMILY_ADDRESS_SIZE(route->family));
+                                if (r < 0)
+                                        return log_link_error_errno(link, r, "Could not append RTA_GATEWAY attribute: %m");
+                        } else if (route->family == AF_INET6) {
+                                r = rta_add_attribute(rta, RTA_GATEWAY, &route->gw.in6, FAMILY_ADDRESS_SIZE(route->family));
+                                if (r < 0)
+                                        return log_link_error_errno(link, r, "Could not append RTA_GATEWAY attribute: %m");
+                        }
```
Comment (yuwata) [inline src/network/networkd-multipath-route.c]:
```
if (route->family == AF_INET)
    r = rta_add_attribute()
else if (route->family == AF_INET6)
    r = rta_add_attribute()
if (r < 0)
    return log_link_error_errno();
```

File: src/network/networkd-multipath-route.c [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+
+                rtnh = RTNH_NEXT(rtnh);
+        }
+
+        r = sd_netlink_message_append_data(req, RTA_MULTIPATH, RTA_DATA(rta), rta->rta_len);
+        if (r < 0)
+                return log_link_error_errno(link, r, "Could not append RTA_MULTIPATH attribute: %m");
+
+        r = netlink_call_async(link->manager->rtnl, NULL, req, multipath_route_add_handler,
+                                  link_netlink_destroy_callback, link);
```
Comment (yuwata) [inline src/network/networkd-multipath-route.c]: spurious indentation.

File: src/network/networkd-multipath-route.c [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+        if (r < 0)
+                return log_link_error_errno(link, r, "Could not send rtnetlink message: %m");
+
+
```
Comment (yuwata) [inline src/network/networkd-multipath-route.c]: doubled blank line.

File: src/systemd/sd-netlink.h [hunk 7]
```diff
@@ -86,13 +86,16 @@ int sd_netlink_message_append_u16(sd_netlink_message *m, unsigned short type, ui
 int sd_netlink_message_append_u32(sd_netlink_message *m, unsigned short type, uint32_t data);
 int sd_netlink_message_append_u64(sd_netlink_message *m, unsigned short type, uint64_t data);
 int sd_netlink_message_append_data(sd_netlink_message *m, unsigned short type, const void *data, size_t len);
+int sd_netlink_message_append_attribute(sd_netlink_message *m, unsigned short type);
```
Comment (yuwata) [inline src/systemd/sd-netlink.h]: Please drop this.

File: src/systemd/sd-netlink.h [hunk 8]
```diff
@@ -86,13 +86,16 @@ int sd_netlink_message_append_u16(sd_netlink_message *m, unsigned short type, ui
...
 int sd_netlink_message_append_in_addr(sd_netlink_message *m, unsigned short type, const struct in_addr *data);
 int sd_netlink_message_append_in6_addr(sd_netlink_message *m, unsigned short type, const struct in6_addr *data);
 int sd_netlink_message_append_sockaddr_in(sd_netlink_message *m, unsigned short type, const struct sockaddr_in *data);
 int sd_netlink_message_append_sockaddr_in6(sd_netlink_message *m, unsigned short type, const struct sockaddr_in6 *data);
 int sd_netlink_message_append_ether_addr(sd_netlink_message *m, unsigned short type, const struct ether_addr *data);
 int sd_netlink_message_append_cache_info(sd_netlink_message *m, unsigned short type, const struct ifa_cacheinfo *info);
 
+int rta_add_attribute(struct rtattr *rta, unsigned short type, const void *data, size_t data_length);
```
Comment (yuwata) [inline src/systemd/sd-netlink.h]: Please move this to netlink-util.h

**Comment by yuwata** [comment]
Hmm, by using the `[MultiPathRoute]` section, we cannot create "multiple" multi-path route. I mean the following is not achievable.
```
ip route add 192.168.0.1 dev dummy99 nexthop via 192.168.11.1 dev inet.11 weight 5 nexthop via 192.168.22.2 dev inet.22 weight 10
ip route add 192.168.0.2 dev dummy99 nexthop via 192.168.11.1 dev inet.11 weight 5 nexthop via 192.168.22.2 dev inet.22 weight 10
```
How about using the following inlined setting?
```
[Route]
NextHop=192.168.11.1 inet.11 5
NextHop=192.168.22.2 inet.11 10
```

Note that, I guess the reason why the current code does not work is that message does not contain `RTA_DST`.

**Comment by ssahani** [comment]
That is ok but does this work. I have no clue why it does not work on one request. Everything same as iproute. I have no clue what I am missing.

**Pull request closed without merging**